test/testTrainingFunction.php

```diff
@@ -0,0 +1,162 @@
+                'difficulty'=>'hard',
+                'weight_added_kg'=>50,
+                'comments'=>'blah blah blah',
+                'html'=>"<div class='exercise_container'><h3> Max Hang <span class='weight_added_span'> | 50kg Added </span></h3><p>blah blah blah</p></div>"
+            ],
+            [
+                'date'=>'2022-04-06',
+                'exercise'=>'Repeater',
+                'difficulty'=>'easy',
+                'weight_added_kg'=>10,
+                'comments'=>'these are some words',
+                'html'=>"<div class='exercise_container'><h3> Repeater <span class='weight_added_span'> | 10kg Added </span></h3><p>these are some words</p></div>"
+            ]
+        ];
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testAddHtmlToWorkoutsFailure_1(){
+        $db_array = [
+            [
+                'exercise'=>'Max Hang',
+                'difficulty'=>'hard',
+                'weight_added_kg'=>50,
+                'comments'=>'blah blah blah',
+            ],
+            [
+                'exercise'=>'Repeater',
+                'difficulty'=>'easy',
+                'weight_added_kg'=>10,
+                'comments'=>'these are some words',
+            ]
+        ];
+        $result = addHtmlToWorkouts($db_array);
+        $expected = [];
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testPerDateOutputSuccess_1(){
+        $db_array = [
+            [
+                'date'=>'2022-07-10',
+                'exercise'=>'Max Hang',
+                'difficulty'=>'hard',
+                'weight_added_kg'=>50,
+                'comments'=>'blah blah blah',
+                'html'=>"<div class='exercise_container'><h3> Max Hang <span class='weight_added_span'> | 50kg Added </span></h3><p>blah blah blah</p></div>"
+            ],
+            [
+                'date'=>'2022-04-06',
+                'exercise'=>'Repeater',
+                'difficulty'=>'easy',
+                'weight_added_kg'=>10,
+                'comments'=>'these are some words',
+                'html'=>"<div class='exercise_container'><h3> Repeater <span class='weight_added_span'> | 10kg Added </span></h3><p>these are some words</p></div>"
+            ]
+        ];
+        $date_array = [
+            '2022-07-10',
+            '2022-04-06'
+        ];
+        $result = perDateOutput($db_array, $date_array);
+        $expected = "<div class='date_container'><h2 class='date_text'>2022-07-10</h2><div class='exercise_container'><h3> Max Hang <span class='weight_added_span'> | 50kg Added </span></h3><p>blah blah blah</p></div></div><div class='date_container'><h2 class='date_text'>2022-04-06</h2><div class='exercise_container'><h3> Repeater <span class='weight_added_span'> | 10kg Added </span></h3><p>these are some words</p></div></div>";
+        $this->assertEquals($expected, $result);
+    }
+
+    public function testPerDateOutputFailure_1(){
```

Comment:
no need for the `_1` on any of the test functions

Reply:
also again i would add a comment explaining what the failure and graceful handling are